### PR TITLE
Replace for PHP tmpfile function

### PR DESCRIPTION
**Pull request by unix4you2**
Some shared hostings doesn't allow to use the tmpfile() PHP functions (i.e. when this function is disabled via  disable_functions  in the php.ini), **so Adminer stop working in that servers**.  This change create a random name for a temporary file that use the standard fopen function in that cases.

### CODE REVIEW

**Comment by gerryd** [comment]
Hi John,

Thanks for the pull request!

I would like to understand the problem properly before including it in the project. I am aware some shared hosting services do indeed block certain PHP functions. I was wondering if you could perhaps enlighten me as to which functions, exactly, are blocked in your case. For example, I am wondering if calling `tempnam()` instead would be a solution?

**Comment by unix4you2** [comment]
Hi @gerryd 

We operate in an hybrid cloud and our software runs in many places or kinds of servers.   When you are in an onpremise server (or cloud server) that you control then you could use the tmpfile function (unless you want to lock it for security purposes).

But in some cases, developers and admins are installing a copy of Adminer under servers that the don't control the config files for PHP.   So when they try to run Adminer the site goes down because PHP language says that there is no a tmpfile function (A syntax error that stop all the script execution).

The patch I wrote do two things:

1. Evaluate if the tmpfile function is recognized by the php in the server and then it works like all the time.  If you check the standar line doesnt have changes and return the pointer to the file in te constructor.
2. In case that tmpfile was not recognized then use a standar fwrite function (that its normally enabled in that kind of servers) to create a new tmpfile "manually" with permissions w+ to be used by adminer and return the pointer.

Today we have around 150+ installations of adminer running by www.practico.org so we can check it in many environments and engines.    This pull request was sent to the adminer origianl repo but as you saw Vrana is not working more in the tool.

Regards.

**Comment by LionelLaffineur** [comment]
Hi John,

Can you explain shortly how you reach that functionality ?  I'm looking in the code and see that it's used to generate name of a tar file, but can't find out how to dump as tar.

Regards,
L.

**Comment by unix4you2** [comment]
Hi @LionelLaffineur 

It doesn't create a tar!  This patch helps to implement Adminer in servers where the admin user disabled the tmpfile function.

Please check the diff in [tmpfile.inc.php](https://github.com/adminerevo/adminerevo/pull/18/files).   At your left the original version that **uses only tmpfile** function.  But at your right you can see the enhacenment that check first if the function is available and if not, then create a random POINTER to a new file with w+ permissions (that is the same effect that the use of [tmpfile](https://www.php.net/manual/en/function.tmpfile.php) that return a resource to a file but using [fopen](https://www.php.net/manual/es/function.fopen.php), that return  the same class of resource.

So you need a pointer to the resource, not a tar or specific file because php just need a file to write the content that Adminer wants.

¿Why is that necessary?  Just because there are a lot of hosting providers and administrators that use the directive [**_disable_functions =_**](https://www.php.net/manual/en/ini.core.php#ini.disable-functions) in their PHP config to disallow the use of tmpfile function.   So the Adminer stop working in those kind of systems.   As in other messages that I sent in this new repo, that patch was created because many of our customers need to deploy in systems that have restrictions but Jakub Vrána didn't accept the pull request because he is not working anymore in the project.

Feel free to contact me at (unix4you2 at gmail dot com) to talk about or to make a conference to explain to you that.

Best regards.

image

**Comment by LionelLaffineur** [comment]
My question is more to know how I could simulate calling that function ?

**Comment by zzjin** [comment]
After reading the diff, 
the main difference is that `tmpfile()` function returns a handler witch file was created under '/tmp/' folder, @see [php manual](https://www.php.net/manual/en/function.tmpfile.php#122678),
and this pr create a random-named file at adminer's script same location and return file handler.

IMO, if some `hosting providers` just disable create tmp file under tmp folder via `disable_functions=tmpfile` for safety purpose (or some kind like this), but allow users to create any file with adminer script same folder.

**Witch seems not like safer.**

Maybe we can change it to create a in-memory tmp file by using `fopen("php://memory", "w+")`?

**Comment by unix4you2** [comment]
Hi there. I think the same.
At the moment I solved the problem for that kind of hostings and servers using that tmp file alternative but as @zzjin says it could be more secure if we use a php://memory resource because the file will be auto-destroyed from server after it's execution (it's the same effect that the __destruct() in my first version)
Regards.

**Comment by Roy-Orbison** [comment]
I don't think this is really the job of Adminer. Providing a crude polyfill that may not suit all scenarios could cause more problems than it solves. In Adminer, the function is used for dumping data, so writing large datasets to the wrong place could quickly exhaust resources or compromise security.

_However_, the error condition of the missing/disabled function should definitely be handled better. It wrongly assumes the function will always exist and never return `false`, but nor should it assume it knows the best way to work around the issue for every environment. If you have such common functions disabled, the effects are not limited to Adminer, so it should be resolved for all PHP code running there.

TL;DR: Adminer should tell you if `tmpfile()` is missing but not dictate how to fix it.
